packages/@react-aria/utils/src/domHelpers.ts

```diff
@@ -9,6 +9,49 @@ export const getOwnerWindow = (
     return el;
   }
 
-  const doc = getOwnerDocument(el as Element | null | undefined);
-  return doc.defaultView || window;
+  const doc = getRootNode(el as Element | null | undefined);
+  return doc instanceof ShadowRoot ?  doc.ownerDocument.defaultView || window :  doc?.defaultView || window;
+};
+
+export const getRootNode = (el: Element | null | undefined): Document | ShadowRoot | null => {
+  if (!el) {
+    // Return the main document if the element is null or undefined
+    return document;
+  }
+
+  const rootNode = el.getRootNode ? el.getRootNode() : document;
+
+  // If the element is disconnected, return null
+  if (!el.isConnected) {
+    return null;
+  }
+
+  // Return the root node, which can be either a Document or a ShadowRoot
+  return rootNode instanceof Document ? rootNode : rootNode as ShadowRoot;
```

Comment:
Typescript will complain as `getRootNode` returns `Node` that will change based on where it is called. I updated the snippet to use type guards instead of casting.